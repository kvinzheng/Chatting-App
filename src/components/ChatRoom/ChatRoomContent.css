.ChatRoomContent {
  width: 100%;
  height: 100%;

  display: flex;
  flex-direction: column;
  background-color: #e0e0e0;
}
.ChatRoomContent-messages {
  width: 100%;
  height: 100%;

  background-color: #e0e0e0;
  display: flex;
  flex-direction: column;
  overflow: scroll;
}

.ChatRoomContent-inputBoxWrapper {
  background-color: white;
  padding: 25px 35px;
  display: flex;
}

.ChatRoomContent-submitButton {
  background-color: white;
  border: none;
  color: #6495ed;
  font-size: 17px;
}
.ChatRoomContent-submitButton:hover {
  cursor: pointer;
}
