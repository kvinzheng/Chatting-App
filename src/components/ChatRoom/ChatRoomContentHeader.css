.ChatRoomContentHeader {
  width: 100%;
  padding: 26px 0px;
  display: flex;
  align-items: center;
  justify-content: center;
  background: white;
  color: #585858;
  flex-direction: column;

  margin-bottom: 10px;
  -moz-box-shadow: 0 4px 4px rgba(0, 0, 0, 0.4);
  -webkit-box-shadow: 0 4px 4px rgba(0, 0, 0, 0.4);
  box-shadow: 0 4px 4px rgba(0, 0, 0, 0.4);
}

.ChatRoomContentHeader-title {
  margin-bottom: 10px;
}

.ChatRoomContentHeader-current {
  color: red;
}
