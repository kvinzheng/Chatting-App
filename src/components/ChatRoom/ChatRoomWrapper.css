.ChatRoomWrapper {
  display: inline-block;
  width: 70%;
}
