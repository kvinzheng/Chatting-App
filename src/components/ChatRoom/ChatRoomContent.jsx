/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState, useRef } from "react";
import { connect } from "react-redux";
import socketIOClient from "socket.io-client";
import PropTypes from "prop-types";

import InputBox from "../../common/InputBox";
import Message from "../../common/Message";
import ChatRoomContentHeader from "./ChatRoomContentHeader";
import { fetchSingleRoom, updateActiveUsers } from "../../actions/room";
import {
  postMessageSingleRoom,
  fetchSingleRoomMessages,
} from "../../actions/message";

import "./ChatRoomContent.css";

import { SCROLL_OPTIONS, NEW_CHAT_MESSAGE, SERVER_HOST } from "../../consts";

export const ChatRoomContent = ({
  currentUserName,
  postMessageSingleRoom,
  fetchSingleRoom,
  fetchSingleRoomMessages,
  messagesDetail,
  activeRoom,
  updateActiveUsers,
}) => {
  const [socket, setSocket] = useState(null);
  const [inputValue, setInputValue] = useState("");
  const inputBoxRef = useRef();
  const activeRoomId = activeRoom.id || 0;
  const messagesList = messagesDetail[activeRoomId] || [];

  const createSocketIO = () => {
    const socket = socketIOClient(SERVER_HOST, {
      query: { roomId: activeRoomId, user: currentUserName },
    });
    socket.on(NEW_CHAT_MESSAGE, ({ message, room }) => {
      const newactiveUsers = room.users;
      const validUserCheck =
        newactiveUsers &&
        activeRoom.users &&
        newactiveUsers.join("/") !== activeRoom.users.join("/");

      if (validUserCheck) {
        updateActiveUsers(newactiveUsers);
      }
      postMessageSingleRoom(activeRoomId, message);
    });
    return socket;
  };

  //create socket when id change
  useEffect(() => {
    const fetApi = async () => {
      await Promise.all([
        fetchSingleRoom(activeRoomId),
        fetchSingleRoomMessages(activeRoomId),
      ]);
    };
    fetApi();

    const newSocket = createSocketIO();
    setSocket(newSocket);
    return () => newSocket.disconnect();
  }, [activeRoomId]);

  //scroll to the inputbox
  useEffect(() => {
    inputBoxRef.current.scrollIntoView(SCROLL_OPTIONS);
  }, [messagesList.length]);

  const handleMessageSubmit = (e) => {
    e.preventDefault();
    socket.emit(NEW_CHAT_MESSAGE, inputValue);
    setInputValue("");
  };

  return (
    <div className="ChatRoomContent">
      <ChatRoomContentHeader
        currentUserName={currentUserName}
        userNamesList={activeRoom.users}
        name={activeRoom.name}
      />

      <div className="ChatRoomContent-messages">
        {messagesList.map(({ name, message }, index) => {
          return (
            <Message
              myself={name === currentUserName}
              author={name}
              key={index}
            >
              {message}
            </Message>
          );
        })}
        <div className="ChatRoomScroll" ref={inputBoxRef} />
      </div>
      <form
        className="ChatRoomContent-inputBoxWrapper"
        onSubmit={handleMessageSubmit}
      >
        <InputBox
          placeholder="Type a message..."
          name="message-input"
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
        />
        <button
          type="link"
          className="ChatRoomContent-submitButton"
          onClick={handleMessageSubmit}
        >
          Send
        </button>
      </form>
    </div>
  );
};
export const mapStateToProps = (state) => {
  return {
    messagesDetail: state.messages,
    rooms: state.rooms.data,
    activeRoom: state.rooms.active,
  };
};

ChatRoomContent.propTypes = {
  messagesDetail: PropTypes.object,
  rooms: PropTypes.array,
  activeRoom: PropTypes.object,
  updateActiveUsers: PropTypes.func,
};
export default connect(mapStateToProps, {
  postMessageSingleRoom,
  fetchSingleRoom,
  fetchSingleRoomMessages,
  updateActiveUsers,
})(ChatRoomContent);
