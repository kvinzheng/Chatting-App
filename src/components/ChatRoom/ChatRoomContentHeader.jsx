import React from "react";
import "./ChatRoomContentHeader.css";
import PropTypes from "prop-types";

const ChatRoomContentHeader = ({ currentUserName, userNamesList, name }) => {
  return (
    <div className="ChatRoomContentHeader">
      <h1 className="ChatRoomContentHeader-title">{name}</h1>
      <div>
        <span className="ChatRoomContentHeader-current">
          {currentUserName} ,
        </span>

        <span>
          {userNamesList &&
            userNamesList.filter((name) => name !== currentUserName).join(", ")}
        </span>
      </div>
    </div>
  );
};

ChatRoomContentHeader.propTypes = {
  currentUserName: PropTypes.string,
  userNamesList: PropTypes.array,
  name: PropTypes.string,
};
export default ChatRoomContentHeader;
