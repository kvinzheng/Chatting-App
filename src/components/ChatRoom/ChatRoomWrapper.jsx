/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import { fetchRooms } from "../../actions/room";
import ChatRoomContent from "./ChatRoomContent";
import PropTypes from "prop-types";

import "./ChatRoomWrapper.css";

const ChatRoomWrapper = ({ history, currentUser, fetchRooms }) => {
  if (!currentUser.login) {
    history.push("/");
  }
  useEffect(() => {
    fetchRooms();
  }, []);

  return (
    <div className="ChatRoomWrapper">
      <ChatRoomContent currentUserName={currentUser.username} />
    </div>
  );
};

export const mapStateToProps = (state) => {
  return {
    currentUser: state.users,
  };
};

ChatRoomWrapper.propTypes = {
  currentUser: PropTypes.object,
  fetchRooms: PropTypes.func,
  history: PropTypes.object,
};

export default connect(mapStateToProps, { fetchRooms })(
  withRouter(ChatRoomWrapper)
);
