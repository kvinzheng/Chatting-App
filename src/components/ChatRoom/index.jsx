import React from "react";
import LeftNav from "../LeftNav/index";
import ChatRoomWrapper from "../ChatRoom/ChatRoomWrapper"
import "./index.css";


const ChatRoom = () => (
  <div className="ChatRoom">
    <LeftNav />
    <ChatRoomWrapper />
  </div>
);

export default ChatRoom;
