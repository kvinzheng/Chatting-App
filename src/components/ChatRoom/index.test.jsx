import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import ChatRoom from "./index";

configure({ adapter: new Adapter() });

describe("ChatRoom Component", () => {
  it("should render without any data defined", () => {
    const component = shallow(<ChatRoom />);
    expect(component).toMatchSnapshot();
  });
});
