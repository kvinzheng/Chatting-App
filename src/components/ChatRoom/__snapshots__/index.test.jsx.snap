// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoom Component should render without any data defined 1`] = `ShallowWrapper {}`;
